queue/queue.go

```diff
@@ -0,0 +1,128 @@
+package queue
+
+import (
+	"context"
+	"encoding/binary"
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/store"
+)
+
+// Put queues the message in the database to be executed at given time.
+// Due to the implementation details, message is guaranteed to be executed
+// after given time, but not exactly at given time.
+func Put(db weave.KVStore, runAt time.Time, msg weave.Msg) error {
+	msgb, err := msg.Marshal()
+	if err != nil {
+		return errors.Wrap(err, "marshal message")
+	}
+
+	for {
+		key := queueKey(runAt)
+		if ok, err := db.Has(key); err != nil {
+			return errors.Wrap(err, "cannot check key existance")
+		} else if ok {
+			// If the key is already in use, instead of storing a
+			// list of messages under each key, which is a very
+			// unlikely to happen, increase the execution time by
+			// the smallest duration.
+			// Message is guaranteed to be executed not earlier
+			// than given time, NOT at exactly given time.
+			runAt = runAt.Add(1)
+			continue
+		}
+
+		if err := db.Set(key, msgb); err != nil {
+			return errors.Wrap(err, "cannot update queue")
+		}
+		return nil
+	}
+}
+
+// Pop removes from the queue a single message that reached its execution time
+// and returns it. It returns ErrEmpty if there is no message suitable for
+// processing.
+func Pop(db weave.KVStore, now time.Time) (weave.Msg, error) {
+	// since := queueKey(time.Time{}) // Zero time is early enough.
+	// until := queueKey(now)
+	// it := db.Iterate(since, until)
+	// ...
+
+	// TODO: iterate from 0 a.d. till now and return the first message
+	// encountered. Return ErrEmpty if no message was found.
+	panic("not implemented")
+}
+
+func queueKey(t time.Time) []byte {
+	rawTime := make([]byte, 8)
+	binary.BigEndian.PutUint64(rawTime, uint64(t.UnixNano()))
```

Comment:
Nanosecond? We could round off to second.

That limits how many messages can be processed per block to 1/second of block time. Which is actually a nice feature in my opinion.

Either is fine, just comment on the precision

Reply:
I have added the comment and changed the granularity to second in https://github.com/iov-one/weave/pull/837/commits/3c31063e01a274ac813ff8f17fceff5588edb2a9

Reply:
You mentioned needing some reliable id to identify these issues. Does it need to be a primary key with an index based on time? Or can we just add some arbitrary "id" label to the payload (not just Msg)?

Let us think a bit more on this one. You also realize

```go
var msg Msg
msg.Unmarshal(data)
```

will always fail, as we need a concrete instance... which means we need some oneof casting here. (This is the reason go-amino allows runtime registration on interfaces. It does make coding in go easier, but at the cost of an undocumented and unmaintainable api for the client)